pkg/istio-agent/xds_proxy.go

```diff
@@ -708,3 +721,165 @@ func sendWithTimeout(ctx context.Context, sendFunc func(errorChan chan error)) e
+			break
+		}
+	}
+
+	p.tapMutex.Unlock()
+
+	if !found {
+		return fmt.Errorf("tap request not found")
+	}
+	return nil
+}
+
+// directResponseToTapper() forwards an Istiod debugging discovery response to a waiting channel
+func (p *XdsProxy) directResponseToTapper(resp *discovery.DiscoveryResponse) {
+	p.tapMutex.Lock()
+	requestsQueue, ok := p.tapRequests[resp.TypeUrl]
+	if ok {
+		requestChan := requestsQueue[0]
+		requestsQueue = requestsQueue[1:]
+		p.tapRequests[resp.TypeUrl] = requestsQueue
+		requestChan <- resp
+	} else {
+		log.Errorf("Istiod debug response %q arrived unexpectedly", resp.TypeUrl)
+	}
+	p.tapMutex.Unlock()
+}
+
+func (p *XdsProxy) makeTapHandler() func(w http.ResponseWriter, req *http.Request) {
+	return func(w http.ResponseWriter, req *http.Request) {
+		serviceAccount := env.RegisterStringVar("SERVICE_ACCOUNT", "default", "").Get()
+		podNamespace := env.RegisterStringVar("POD_NAMESPACE", "default", "").Get()
+		istioVersion := env.RegisterStringVar("ISTIO_META_ISTIO_VERSION", "1.9.1", "").Get()
+
+		typeURL := fmt.Sprintf("istio.io%s", req.URL.Path)
+		responseChan, err := p.tapRequest(&discovery.DiscoveryRequest{
+			Node: &envoy_v3.Node{
+				Id: "sidecar~1.1.1.1~debug~cluster.local", // TODO pick up if we have it
+				Metadata: &structpb.Struct{
+					Fields: map[string]*structpb.Value{
+						"SERVICE_ACCOUNT": {Kind: &structpb.Value_StringValue{StringValue: serviceAccount}},
+						"NAMESPACE":       {Kind: &structpb.Value_StringValue{StringValue: podNamespace}},
+						"GENERATOR":       {Kind: &structpb.Value_StringValue{StringValue: "event"}},
+						"ISTIO_VERSION":   {Kind: &structpb.Value_StringValue{StringValue: istioVersion}},
+					},
+				},
+			},
+			TypeUrl: typeURL,
+		})
+		if err != nil {
+			w.WriteHeader(503)
+			fmt.Fprintf(w, "%v\n", err)
+			return
+		}
+
+		// Timeout if no response from Istiod after 5 seconds
+		go func() {
+			time.Sleep(5 * time.Second)
+			responseChan <- nil
+		}()
+
+		// Wait for Istiod to respond
+		response := <-responseChan
+
+		if response == nil {
+			log.Infof("timed out waiting for Istiod to respond to %q", typeURL)
+			err = p.abortTap(typeURL, responseChan)
+			if err != nil {
+				log.Warnf("failed to cleanup tap for %q: %v", typeURL, err)
+			}
+			w.WriteHeader(504)
+			return
+		}
+
+		// Send Istiod's response as HTTP body
+		j, err := json.Marshal(response)
+		if err != nil {
+			w.WriteHeader(500)
+			fmt.Fprintf(w, "%v\n", err)
+		}
+
+		w.Header().Add("Content-Type", "application/json")
+		_, err = w.Write(j)
+		if err != nil {
+			log.Infof("fail to write debug response: %v", err)
+		}
+	}
+}
+
+// initDebugInterface() listens on :15009 for requests with path /debug/... forwards the paths as Istiod XDS requests
+func (p *XdsProxy) initDebugInterface() error {
+	p.tapRequests = make(map[string][]chan *discovery.DiscoveryResponse)
+
+	httpMux := http.NewServeMux()
+	httpMux.HandleFunc("/debug/", p.makeTapHandler())
+
+	HTTPAddr := ":15009"
```

Comment:
I know this was for demo, but we cannot listen on 0.0.0.0. Either UDS or localhost.